src/c/device-coap.c

```diff
@@ -0,0 +1,225 @@
+
+  driver->lc = lc;
+  driver->security_mode = find_security_mode (iot_data_string_map_get_string (config, SECURITY_MODE_KEY));
+
+  if (driver->security_mode == SECURITY_MODE_PSK)
+  {
+    const char *conf_psk_key = iot_data_string_map_get_string (config, PSK_KEY_KEY);
+    if (strlen (conf_psk_key))
+    {
+      iot_data_t *key_array = iot_data_alloc_array_from_base64 (conf_psk_key);
+      driver->psk_key = key_array;
+      iot_log_info (lc, "PSK key len %u", iot_data_array_length (key_array));
+    }
+    else
+    {
+      iot_log_error (lc, "PSK key not in configuration");
+      driver->psk_key = NULL;
+      return false;
+    }
+  }
+  else
+  {
+    driver->psk_key = NULL;
+  }
+
+  /* CoAP server bind address as text */
+  const char *bind_addr = iot_data_string_map_get_string (config, COAP_BIND_ADDR_KEY);
+  if (bind_addr)
+  {
+    driver->coap_bind_addr = iot_data_alloc_string (bind_addr, IOT_DATA_COPY);
+  }
+  else
+  {
+    iot_log_error (lc, "CoAP bind address not in configuration");
+    driver->coap_bind_addr = NULL;
+    return false;
+  }
+
+  iot_log_debug (lc, "Init complete");
+  return true;
+}
+
+static bool coap_get_handler
+(
+  void *impl,
+  const char *devname,
+  const devsdk_protocols *protocols,
+  uint32_t nreadings,
+  const devsdk_commandrequest *requests,
+  devsdk_commandresult *readings,
+  const devsdk_nvpairs *qparms,
+  iot_data_t **exception
+)
+{
+  (void) impl;
+  (void) devname;
+  (void) protocols;
+  (void) nreadings;
+  (void) requests;
+  (void) readings;
+  (void) qparms;
+  (void) exception;
+  return true;
+}
+
+static bool coap_put_handler
+(
+  void *impl,
+  const char *devname,
+  const devsdk_protocols *protocols,
+  uint32_t nvalues,
+  const devsdk_commandrequest *requests,
+  const iot_data_t *values[],
+  iot_data_t **exception
+)
+{
+  (void) impl;
+  (void) devname;
+  (void) protocols;
+  (void) nvalues;
+  (void) requests;
+  (void) values;
+  (void) exception;
+  return true;
+}
+
+static void coap_stop (void *impl, bool force) {}
+
+int main (int argc, char *argv[])
+{
+  coap_driver * impl = malloc (sizeof (coap_driver));
+  memset (impl, 0, sizeof (coap_driver));
+
+  devsdk_error e;
+  e.code = 0;
+
+  /* Device Callbacks */
+  devsdk_callbacks coapImpls =
+  {
+    coap_init,
+    NULL,              /* discovery */
+    coap_get_handler,
+    coap_put_handler,
+    coap_stop,
+    NULL,              /* auto-event started */
+    NULL,              /* auto-event stopped */
+    NULL,              /* device added */
+    NULL,              /* device updated */
+    NULL               /* device removed */
+  };
+
+  /* Initialize a new device service */
+  devsdk_service_t *service = devsdk_service_new
+    ("device-coap", VERSION, impl, coapImpls, &argc, argv, &e);
+  ERR_CHECK (e);
+  impl->service = service;
+
+  int res = 0;
+  int n = 1;
+  while (n < argc)
+  {
+    if (strcmp (argv[n], "-h") == 0 || strcmp (argv[n], "--help") == 0)
```

Comment:
Thanks for flagging. This code was speculative based on what I saw in device-bacnet-c. However, device-bacnet-c no longer processes these options, so I will remove here too.

Reply:
To correct my previous comment, the other arguments are handled in `devsdk_service_new()`. However, based on device-bacnet-c and device-grove-c, the idiomatic way to display the other arguments is via `devsdk_usage()`. That approach definitely is cleaner since the options are handled in the device SDK code as well. I will update to use that function.

Reply:
Addressed in f5e9fe5